src/kernel/keyboard.zig

```diff
@@ -0,0 +1,353 @@
+
+        comptime var i = 0;
+        inline while (i < QUEUE_SIZE - 1) : (i += 1) {
+            testing.expectEqual(keyboard.writeKey(.{
+                .position = @intToEnum(KeyPosition, i),
+                .released = false,
+            }), true);
+            testing.expectEqual(keyboard.queue[i].position, @intToEnum(KeyPosition, i));
+            testing.expectEqual(keyboard.queue_end, i + 1);
+            testing.expectEqual(keyboard.queue_front, 0);
+        }
+
+        testing.expectEqual(keyboard.writeKey(.{
+            .position = @intToEnum(KeyPosition, 33),
+            .released = false,
+        }), false);
+        testing.expect(keyboard.isFull());
+    }
+
+    test "readKey" {
+        var keyboard = Keyboard.init();
+
+        comptime var i = 0;
+        inline while (i < QUEUE_SIZE - 1) : (i += 1) {
+            testing.expectEqual(keyboard.writeKey(.{
+                .position = @intToEnum(KeyPosition, i),
+                .released = false,
+            }), true);
+        }
+
+        i = 0;
+        inline while (i < QUEUE_SIZE - 1) : (i += 1) {
+            testing.expectEqual(keyboard.readKey().?.position, @intToEnum(KeyPosition, i));
+            testing.expectEqual(keyboard.queue_end, QUEUE_SIZE - 1);
+            testing.expectEqual(keyboard.queue_front, i + 1);
+        }
+
+        testing.expect(keyboard.isEmpty());
+        testing.expectEqual(keyboard.readKey(), null);
+    }
+};
+
+/// The registered keyboards
+var keyboards: ArrayList(*Keyboard) = undefined;
+
+///
+/// Get the keyboard associated with an ID
+///
+/// Arguments:
+///     id: usize - The ID of the keyboard to get
+///
+/// Return: ?*Keyboard
+///     The keyboard associated with the ID, or null if there isn't one
+///
+pub fn getKeyboard(id: usize) ?*Keyboard {
+    if (keyboards.items.len <= id) {
+        return null;
+    }
+    return keyboards.items[id];
+}
+
+///
+/// Add a keyboard to list of known keyboards
+///
+/// Arguments:
+///     kb: *Keyboard - The keyboard to add
+///
+/// Error: std.mem.Allocator.Error
+///     Allocator.Error.OutOfMemory - Adding the keyboard to the list failed due to there not being enough memory free
+///
+pub fn addKeyboard(kb: *Keyboard) std.mem.Allocator.Error!void {
+    try keyboards.append(kb);
+}
+
+///
+/// Initialise the keyboard system and the architecture's keyboard
+///
+/// Arguments:
+///     allocator: *std.mem.Allocator - The allocator to initialise the keyboard list and architecture keyboard with
+///
+/// Return: ?*Keyboard
+///     The architecture keyboard found, else null if one wasn't detected
+///
+/// Error: std.mem.Allocator.Error
+///     Allocator.Error.OutOfMemory - There wasn't enough memory to initialise the keyboard list or the architecture keyboard
+///
+pub fn init(allocator: *std.mem.Allocator) std.mem.Allocator.Error!?*Keyboard {
+    keyboards = ArrayList(*Keyboard).init(allocator);
+    return try arch.initKeyboard(allocator);
```

Comment:
You don't need the `try` in the return but up to you.

Reply:
Oh yeah :)